source_code/main_mcu/src/LOGIC/logic_totp.c

```diff
@@ -0,0 +1,174 @@
+                    decoded_key[decoded_index] |= j << (shift_offset - LOGIC_TOTP_BASE32_BITS);
+                    shift_offset -= LOGIC_TOTP_BASE32_BITS;
+                }
+                else
+                {
+                    // In this case, we have to split up the 5 bits of the index between the remaining bits of decoded_key[decoded_index]
+                    // and the first bits of decoded_key[decoded_index + 1]. Therefore we first have to shift the index by 5 - shift_offset to the right.
+                    // That means, we have shift_offset (which is also the number of remaining bits in the current byte) bits from the index remaining to put into the current byte.
+                    // Then we need to calculate the new shift_offset, which is basically the old shift_offset + 3 (As already described, the shift_offset is also
+                    // the number of remaining bits in the byte. As we have in the first step used this number of bits from the index, there are 5 - shift_offset bits from the
+                    // index remaining, that have to be put in the next byte. To get the number places the remaining bits have to be shifted in the next byte, we have to subtract
+                    // shift_offset from the number of bits in a byte. If we rearrange the order in that term, we get shift_offset = shift_offset + 8 - 5 = shift_offset + 3)
+                    // As an example for the second character, shift_offset is 3. So we take the first 3 of the 5 index bits and put them into current byte. Then we have 2 bits remaining.
+                    // Those bits need to be shifted then by 6 places to the left, to be at the beginning of the next byte.
+                    decoded_key[decoded_index] |= j >> (LOGIC_TOTP_BASE32_BITS - shift_offset);
+                    shift_offset += LOGIC_TOTP_BITS_PER_BYTE - LOGIC_TOTP_BASE32_BITS;
+                    decoded_index++;
+                    decoded_key[decoded_index] |= j << (shift_offset);
+                }
+            }
+        }
+    }
+}
+
+
+/*! \fn     logic_totp(char* totp_token, const char* key)
+*   \brief  Calculate TOTP Token using given key
+*   \param  calculated totp_token, char[6] expected
+*   \param  Base32 encoded key that is used to calculate totp_token, char[16] expectd
+*   \return void
+*/
+void logic_totp(char* totp_token, const char* key)
+{
+    char hash[20];
+    char decoded_key[10];
+    int truncated_hash = 0;
+    int offset = 0;
+    int shift_offset = LOGIC_TOTP_BITS_PER_BYTE * 3;
+    int token_idx = LOGIC_TOTP_TOKEN_LEN - 1;
+    time_t timestamp = logic_totp_get_time();
+    // step is the message the hash gets calculated for. TOTP is based on the current unix timestamp. As a TOTP token is valid for 30 seconds, we divide the current timestamp by
+    // 30 before calculating the SHA1 HMAC hash
+    time_t step = utils_switch_endiannes(timestamp/30);
+
```

Comment:
I have yet to signed up for a service with a TOTP token that has a different time step than 30 seconds but it could be possible...not sure if we care enough to support anything configurable?

Reply:
spec is 30 seconds IIRC

Reply:
Yes, basically. This is what the RFC says:

_We RECOMMEND a default time-step size of 30 seconds.  This default value of 30 seconds is selected as a balance between security and usability._

So I _think_ we should be fine just supporting 30 seconds.

Reply:
In case there is a different step size used than 30 seconds, how is that specified? When I add 2FA for a Service, that accepts Google Auth, only the secret is transferred, no step size, as well as no token length different than 6 digits.